src/authentication/authentication.service.ts

```diff
@@ -44,4 +54,54 @@ export class AuthenticationService {
       token: this.generateAuthToken({ email: user.email, sub: user.id }),
     }
   }
+
+  async updatePasswordWithToken(
+    email: string,
+    token: string,
+    password: string,
+  ) {
+    let user
+    try {
+      this.jwtService.verify(token)
+      user = await this.usersService.findUserByResetToken(email, token)
+    } catch (e) {
+      if (
+        e instanceof SyntaxError || // token parse failed
+        e instanceof PrismaClientKnownRequestError // token doesn't match a user
+      ) {
+        throw new BadRequestException('Invalid token')
+      } else if (e instanceof TokenExpiredError) {
+        throw new BadRequestException('Token has expired')
+      }
+
+      throw e
+    }
+
+    try {
+      return await this.usersService.updatePassword(user.id, password, true)
+    } catch (e) {
+      console.log(e)
+      throw new BadRequestException('Failed to update password')
+    }
+  }
+
+  async generatePasswordResetToken(userId: number): Promise<User> {
+    const token = nanoid(48)
+
+    const jwt = this.jwtService.sign(
+      { token },
+      { expiresIn: PASSWORD_RESET_TOKEN_TTL },
+    )
+
+    try {
+      return await this.usersService.setResetToken(userId, jwt)
+    } catch (e) {
+      if (e instanceof PrismaClientKnownRequestError) {
+        console.log('Could not find user to reset password')
+        throw new NotFoundException('User not found')
+      }
+
+      throw e
+    }
+  }
```

Comment:
This method should just generate the token, then up in the `AuthenticationController` up where it's called, it should call `userService.update(...)` to add the token to the user record.

That way, also, the exception handling can be handled there.

Also.. we don't want to reveal to the client whether users exist or don't in our DB.. that's a security vulnerability. We can just respond w/ `204` from the controller as a way of saying, "Ok.. we accepted your request and we'll do something w/ it if it makes sense to do so"

If we want to, before responding we can do a `this.logger.warn('request to reset password for non-existent user'` if it seems valuable to do so.